_posts/2019-12-18-road-to-scala-3.md

```diff
@@ -0,0 +1,174 @@
+
+### What about macros?
+
+While Scala 3 can generally consume libraries built with Scala 2.13, this does not work for macros. Macro methods defined in 2.13 libraries cannot be used in Scala 3, because the Scala 3 compiler cannot execute the macro at compile-time.
+
+The macro system from Scala 2 is deeply tied to the internals of the Scala 2 compiler and cannot be migrated in a compatible fashion to Scala 3. Instead, Scala 3 ships with a [new macro system](https://dotty.epfl.ch/docs/reference/metaprogramming/toc.html). This means that macro definitions have to be re-written when migrating a codebase to Scala 3. Cross-building is possible by having separate source files for Scala 2 and 3 macro definitions.
+
+Libraries that define macros can be made available for both Scala 2.13 and Scala 3 by cross-compiling. 
+
+### What breaking changes are in Scala 3, besides macros?
+
+The Scala 3 compiler supports almost all of the Scala 2 language. For a detailed overview, see [this page](https://dotty.epfl.ch/docs/reference/features-classification.html#dropped-constructs). Language features that have been dropped are [early initializers](https://dotty.epfl.ch/docs/reference/dropped-features/early-initializers.html), [existential types](https://dotty.epfl.ch/docs/reference/dropped-features/existential-types.html), [weak conformance](https://dotty.epfl.ch/docs/reference/dropped-features/weak-conformance.html) and [`DelayedInit`](https://dotty.epfl.ch/docs/reference/dropped-features/delayed-init.html). (Some of these removals are still pending SIP committee approval.)
+
+### What about performance?
+
+There are multiple aspects of performance:
+
+* **Compiler performance**. The performance of the Scala 3 compiler is being monitored continuously to prevent regressions (http://dotty-bench.epfl.ch). We will work on enabling benchmarks that compare the compiler performance between Scala 2.13 and Scala 3.
+
+* **Runtime performance of generated code**. The code generation phase of the Scala 3 compiler is the same as in Scala 2.13, and also the ABI (the way Scala features are encoded in class files) in Scala 3.0 will be the same. We expect that code compiled with Scala 3 will perform the same, and we will work with the community to verify that this is the case. The Scala 2 optimizer will also be available in Scala 3.
+
+* **Performance-oriented language features**. While value classes continue to work, Scala 3 brings [opaque type aliases](https://dotty.epfl.ch/docs/reference/other-new-features/opaques.html) that provide type abstraction without any overhead. [Inline methods](https://dotty.epfl.ch/docs/reference/metaprogramming/inline.html) can be used for methods that are guaranteed to be inlined at compile-time. Scala 3 currently does not implement specialization. We will work with maintainers of performance-sensitive projects to find out what variants of specialization need to be supported. This work will likely land in a Scala 3.x minor release.
```

Comment:
Given that the Scala 2 library uses specialization for Function1 etc. and Tuple2 etc., doesn't this mean that even seemingly innocuous code like

```scala
def foldInts(zero: Int, op: (Int, Int) => Int, array: Array[Int]) = {
  var acc = zero
  var i = 0
  while (i < array.length) { acc = op(acc, array(i)); i += 1 }
  acc
}
```

is going to get like 40x slower due to boxing?

Or is this already fixed?  If it's fixed, now is probably a good time to mention it.

Reply:
We handle function specialization by hand so that one is probably OK.

Reply:
Might be worth mentioning it, then.  Maybe link to any docs on what is handled by hand (just function literals?).  Most uses of specialization are rather esoteric, but this one isn't.

Reply:
Would be great to have this discussion on discourse!